torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,175 @@
+#include "nccl.h"
+#include "torch/csrc/THP.h"
+#include "torch/csrc/Types.h"
+#include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/cuda/THCP.h"
+
+#include <nccl.h>
+#include <sstream>
+#include <unordered_map>
+
+static inline void CHECK(ncclResult_t status)
+{
+  if (status != ncclSuccess) {
+    std::stringstream err;
+    err << "NCCL Error " << status << ": " << ncclGetErrorString(status);
+    throw std::runtime_error(err.str());
+  }
+}
+
+// TODO: make this thread_local + add mutexes
+std::unordered_map<std::string, ncclComm_t*> _communicators;
+
+PyObject * THCPModule_nccl_destroy(PyObject *args) {
+  for(auto it: _communicators) {
+    ncclCommDestroy(*(it.second));
```

Comment:
nit: `it->second`

Reply:
did not work, also not equivalent. 
i am getting `it.second` and dereferencing the result, because I'm getting a ncclComm_t**

Reply:
This looks like it's hiding a bug that you only destroy first communicator in each list